### PR TITLE
Implement more of Numpy Interface

### LINKED ISSUE #29: Implement more of the dunder-ndarray interface.
Implement more of the `ndarray` interface.

# Memory Layout

* [x] [`ndarray.shape`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape)/`np.shape`
* [x]  [`ndarray.ndim`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.ndim.html#numpy.ndarray.ndim)/`np.ndim`
* [x]  [`ndarray.size`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.size.html#numpy.ndarray.size)/`np.size`
* [x]  [`ndarray.nbytes`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.nbytes.html#numpy.ndarray.nbytes)
* [x] [`ndarray.itemsize`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.itemsize.html#numpy.ndarray.itemsize)

# Transposing

* [x] [`transpose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.transpose.html#numpy.transpose)
* [x] [`ndarray.T`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.T.html#numpy.ndarray.T)
* [x] [`swapaxes`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.swapaxes.html#numpy.swapaxes)
* [x] [`rollaxis`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.rollaxis.html#numpy.rollaxis)
* [x] [`moveaxis`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.rollaxis.html#numpy.rollaxis)

# Reshaping
* [x] [`reshape`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html#numpy.reshape)
* [x] [`ravel`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ravel.html#numpy.ravel)
* [x] [`ndarray.flatten`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.flatten.html#numpy.ndarray.flatten)

# GetItem and Friends
* [x] [`ndarray.__getitem__`](https://docs.scipy.org/doc/numpy/user/basics.indexing.html)
* [x] [`broadcast_to`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.broadcast_to.html#numpy.broadcast_to)/`ndarray.broadcast_to`
* [x] [`broadcast_arrays`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.broadcast_arrays.html#numpy.broadcast_arrays)

# Joining Arrays
* [x] [`concatenate`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html#numpy.concatenate)
* [x] [`stack`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.stack.html#numpy.stack)
* [x] [`column_stack`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.column_stack.html#numpy.column_stack)
* [x] [`hstack`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.hstack.html#numpy.hstack)
* [x] [`vstack`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.vstack.html#numpy.vstack)
* [x] [`block`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.block.html#numpy.block)

### PR DESCRIPTION
**Pull request by prasunanand**
Need to add tests.

Fixes: https://github.com/Quansight-Labs/unumpy/issues/29

### CODE REVIEW

**Comment by prasunanand** [comment]
https://github.com/Quansight-Labs/unumpy/pull/30/files#diff-801cbeb7bd729e76ac2c91759b13745dR151

Here Dask implemetation of hstack doesn't follow Numpy API 

```
 def test_functions_coerce(backend, method, args, kwargs):
        backend, types = backend
        try:
            with ua.set_backend(backend, coerce=True):
>               ret = method(*args, **kwargs)

unumpy/tests/test_numpy.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
unumpy/dask_backend.py:80: in __ua_function__
    return getattr(da, method.__name__)(*args, **kwargs)
../../../anaconda3/envs/uarray/lib/python3.7/site-packages/dask/array/routines.py:121: in hstack
    if all(x.ndim == 1 for x in tup):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <tuple_iterator object at 0x7fb717e96590>

>   if all(x.ndim == 1 for x in tup):
E   AttributeError: 'list' object has no attribute 'ndim'

../../../anaconda3/envs/uarray/lib/python3.7/site-packages/dask/array/routines.py:121: AttributeError

```

**Review by hameerabbasi — COMMENTED** [review]
A lot of these could use default implementations that dispatch to the respective property.

**Comment by prasunanand** [comment]
Agreed. Will try to simplify.

**Comment by prasunanand** [comment]
> A lot of these could use default implementations that dispatch to the respective property.

@hameerabbasi I am unable to implement it. Do you have anything in mind regarding solving this ?

**Comment by hameerabbasi** [comment]
You can just add a `default=` kwarg to the `create_numpy` decorator. Look for examples in the `unumpy/_multimethods.py` file.

**Comment by hameerabbasi** [comment]
Defaults are usually available in the docs as well, some just dispatch to properties. And `rollaxis`, etc. can be written in terms of `transpose` if you read the docs and see what they do.

File: unumpy/_multimethods.py [hunk 1]
```diff
@@ -709,3 +709,75 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
...
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def reshape(a, newshape, order="C"):
+    return (a,)
+
+
+@create_numpy(_identity_argreplacer)
+@all_of_type(ndarray)
+def column_stack(tup):
+    return (tup,)
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py:765]:
```suggestion
    return tup
```

File: unumpy/_multimethods.py [hunk 2]
```diff
@@ -709,3 +709,75 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
...
+
+
+@create_numpy(_identity_argreplacer)
+@all_of_type(ndarray)
+def hstack(tup):
+    return (tup,)
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py:771]:
```suggestion
    return tup
```

File: unumpy/_multimethods.py [hunk 3]
```diff
@@ -709,3 +709,75 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
...
+
+
+@create_numpy(_identity_argreplacer)
+@all_of_type(ndarray)
+def vstack(tup):
+    return (tup,)
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py:777]:
```suggestion
    return tup
```

**Review by hameerabbasi — COMMENTED** [review]
Hey. Here's a few fixes. As discussed offline, please add defaults:

* `shape` dispatches to attribute.
* `ndim` dispatches to `len(shape)`
* `size` dispatches to `functools.reduce(operator.mul, shape, 1)`
* `swapaxes`/`rollaxis`/`moveaxis` can be written in terms of `transpose`.
* `reshape` dispatches to attribute.
* `column_stack`/`hstack`/`vstack` dispatch to `concatenate` + `atleast_*d` (which you might need to add).
* block has a list structure, discussed in the call.

**Comment by hameerabbasi** [comment]
Superceded by #32.

**Pull request closed without merging**